src/main/java/com/direwolf20/buildinggadgets/common/building/placement/ConnectedSurface.java

```diff
@@ -0,0 +1,160 @@
+        this.world = world;
+        this.searchingRegion = searchingRegion;
+        this.searching2referenceMapper = searching2referenceMapper;
+        this.searchingCenter = searchingCenter;
+        this.side = side;
+        this.fuzzy = fuzzy;
+    }
+
+    /**
+     * The bounding box of the searchingRegion that is being searched.
+     */
+    @Override
+    public Region getBoundingBox() {
+        return searchingRegion;
+    }
+
+    /**
+     * {@inheritDoc}<br>
+     * <b>inaccurate representation (case 2)</b>:
+     *
+     * @return {@code getBoundingBox().contains(x, y, z)} since it would be costly to check each position
+     */
+    @Override
+    public boolean mayContain(int x, int y, int z) {
+        return searchingRegion.mayContain(x, y, z);
+    }
+
+    /**
+     * @deprecated ConnectedSurface should be immutable, so this is not needed
+     */
+    @Deprecated
+    @Override
+    public IPlacementSequence copy() {
+        return new ConnectedSurface(world, searchingRegion, searching2referenceMapper, searchingCenter, side, fuzzy);
+    }
+
+    /**
+     * @implNote technically BFS fits better with a flood fill algorithm and it simpler to implement with a generator
+     */
+    @Nonnull
+    @Override
+    public Iterator<BlockPos> iterator() {
+        IBlockState selectedBlock = getReferenceFor(searchingCenter);
+
+        return new AbstractIterator<BlockPos>() {
+            private Queue<BlockPos> queue = new ArrayDeque<>(searchingRegion.size());
+            private ObjectSet<BlockPos> searched = new ObjectOpenHashSet<>();
+
+            {
+                queue.add(searchingCenter);
+                searched.add(searchingCenter);
+            }
+
+            @Override
+            protected BlockPos computeNext() {
+                if (queue.isEmpty()) {
+                    return endOfData();
+                }
+
+                //The position is guaranteed to be valid
+                BlockPos current = queue.remove();
+
+                for (int i = -1; i <= 1; i++) {
+                    for (int j = -1; j <= 1; j++) {
+                        BlockPos neighbor = VectorTools.perpendicularSurfaceOffset(current, side, i, j);
+
+                        boolean isSearched = searched.contains(neighbor);
+                        searched.add(neighbor);
+
+                        if (isSearched || !searchingRegion.contains(neighbor) || !isStateValid(selectedBlock, getReferenceFor(neighbor))) {
+                            continue;
+                        }
+
+                        queue.add(neighbor);
+                    }
+                }
+
+                return current;
+            }
+        };
+    }
+
+    private boolean isStateValid(IBlockState filter, IBlockState reference) {
+        boolean isAir = reference.getMaterial() == Material.AIR;
```

Comment:
Technically this would need to be a world.isAirCheck, but as we don't have a world here, I don't think we can do anything about it.

Reply:
`ConnectedSurface#world` is an `IBlockAccess`, which contains an `isAirBlock` method. However for efficiency, since both `ChunkCache` and `World` implementations of `IBlockAccess` just call `isAir` on the retrieved state's block, the same can be done for `reference`:
```
boolean isAir = reference.getBlock().isAir(reference, world, posReference);
```
(where `neighbor` is passed as `posReference` in `ConnectedSurface#iterator`)

Reply:
Although this would make tests invalid -- which would be fine (for now).
Obviously `Block#isAir` rely in a `Block` object, which would crash on its creation, or the constructor call, if we don't have a proper Minecraft runtime context.

Reply:
Regardless of how you handle testing, this is how air checks need to be done when actually using gadgets in-game.

Reply:
This has been marked as resolved, but the `BuildingModes` and `ExchangingModes` classes still have air material checks.